Content.Server/Traits/TraitSystem.Functions.cs

```diff
@@ -444,85 +448,29 @@ public sealed partial class TraitModifyMobState : TraitFunction
-
-    [DataField, AlwaysPushInheritance]
-    public bool? AllowHandInteractWhileDead;
+    [DataField(required: true)]
+    public Dictionary<string, ProtoId<MobStateParametersPrototype>> Params = default!;
 
     public override void OnPlayerSpawn(EntityUid uid,
         IComponentFactory factory,
         IEntityManager entityManager,
         ISerializationManager serializationManager)
     {
-        if (!entityManager.TryGetComponent<MobStateComponent>(uid, out var mobStateComponent))
+        if (!entityManager.TryGetComponent<MobStateComponent>(uid, out var comp))
             return;
 
-        if (AllowMovementWhileCrit is not null)
-            mobStateComponent.AllowMovementWhileCrit = AllowMovementWhileCrit.Value;
-
-        if (AllowMovementWhileSoftCrit is not null)
-            mobStateComponent.AllowHandInteractWhileSoftCrit = AllowMovementWhileSoftCrit.Value;
-
-        if (AllowMovementWhileDead is not null)
-            mobStateComponent.AllowMovementWhileDead = AllowMovementWhileDead.Value;
-
-        if (AllowTalkingWhileCrit is not null)
-            mobStateComponent.AllowTalkingWhileCrit = AllowTalkingWhileCrit.Value;
-
-        if (AllowTalkingWhileSoftCrit is not null)
-            mobStateComponent.AllowTalkingWhileSoftCrit = AllowTalkingWhileSoftCrit.Value;
-
-        if (AllowTalkingWhileDead is not null)
-            mobStateComponent.AllowTalkingWhileDead = AllowTalkingWhileDead.Value;
+        var _reflection = IoCManager.Resolve<IReflectionManager>();
+        var _proto = IoCManager.Resolve<IPrototypeManager>();
 
-        if (DownWhenCrit is not null)
-            mobStateComponent.DownWhenCrit = DownWhenCrit.Value;
-
-        if (DownWhenSoftCrit is not null)
-            mobStateComponent.DownWhenSoftCrit = DownWhenSoftCrit.Value;
-
-        if (DownWhenDead is not null)
-            mobStateComponent.DownWhenDead = DownWhenDead.Value;
-
-        if (AllowHandInteractWhileCrit is not null)
-            mobStateComponent.AllowHandInteractWhileCrit = AllowHandInteractWhileCrit.Value;
-
-        if (AllowHandInteractWhileSoftCrit is not null)
-            mobStateComponent.AllowHandInteractWhileSoftCrit = AllowHandInteractWhileSoftCrit.Value;
-
-        if (AllowHandInteractWhileDead is not null)
-            mobStateComponent.AllowHandInteractWhileDead = AllowHandInteractWhileDead.Value;
+        foreach (var pair in Params) {
+            DebugTools.Assert(_reflection.TryParseEnumReference($"enum.MobState.{pair.Key}", out var e), $"MobState.{pair.Key} does not exist.");
+            MobState state = (MobState) e;
+            MobStateParametersPrototype current = comp.MobStateParams[state];
+            var mod = _proto.Index<MobStateParametersPrototype>(pair.Value);
+            current.MergeWith(mod);
+            //current.FillDefaults();
+        }
+        comp.Dirty(); // why is this deprecated? it's much better than manually resolving entitymanager with ioc
```

Comment:
```suggestion
        comp.Dirty(); // why is this deprecated? it's much better than manually resolving entitymanager with ioc
```
It's deprecated because it's been replaced with Dirty(uid, comp); instead of Dirty(comp). I think you might be able to get around the deprecated warning by doing comp.Dirty(uid);. The reason for the deprecation is because you're supposed to only dirty one component at a time, and if you dirty just the component, the system now has to dirty all of that component, so it needs UID + Component.

However we now have as an engine option, the ability to dirty **individual datafields** in components, feel free to take a look at that if you wish!

Reply:
not sure how its implementation changed over time, but currently `comp.Dirty()` is pretty much equivalent to `Dirty(comp.Owner, comp)`, barring a single resolve of EntityManager. I guess it's marked as obsolete because `Owner` field is obsolete too.
Being able to dirty individual datafields is, i suppose, about custom defined component states? First time i'm hearing about this.